ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -600,28 +627,84 @@ uniformAdversarialChain mbAsc recipe g0 = wrap $ C.createV $ do
-                -- schema
-                hScgWin = C.UnsafeContains (C.windowStart carWin) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                -- first stability window after the intersection in the alternative
-                -- schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- slice of the alternative schema for the first stability
-                -- window after the intersection
-                aScgMv  = C.sliceMV aScgWin mv
-                -- intended number of empty slots in the first stability window
-                -- after the intersection in the alternative schema
-                emptyCountTarget = C.toVar $ S.complementActive S.notInverted (C.windowSize aScgWin) hCount C.+ 1
-            void $ BV.fillInWindow
-                S.inverted
-                (BV.SomeDensityWindow emptyCountTarget (C.windowSize aScgWin))
-                g
-                aScgMv
+    -- | Ensure the density of the adversarial schema is less than the density
+    -- of the honest schema in the first stability window after the intersection
+    -- and in any subwindow that contains the first race to the k+1st block.
+    --
+    -- Ensuring lower density of the subwindows is necessary to avoid chains like
+    --
+    -- > k: 3
+    -- > s: 9
+    -- > H: 0111100
+    -- > A: 1110011
+    --
+    -- where the honest chain wins the race to the k+1st block, might win the
+    -- density comparison if the chains are extended to include a full stability
+    -- window after the intersection, but loses the density comparison if the
+    -- chains aren't extended.
+    --
+    -- For the sake of shrinking test inputs, we also prevent the above scenario
+    -- to occur in any intersection, not just the intersections near the end of
+    -- the chains.
+    --
+    ensureLowerDensityInWindows
+      :: R.StatefulGen sg (ST s) => RI.Race adv -> sg -> C.MVector adv SlotE s S.S -> ST s ()
+    ensureLowerDensityInWindows (RI.Race (C.SomeWindow _ w0)) g mv = do
+        let
+          -- A window after the intersection as short as the shortest of the
+          -- stability window or the first race to the k+1st block.
+          w0' = C.UnsafeContains (C.windowStart w0) (min (C.Count s) (C.windowSize w0))
+          hCount = BV.countActivesInV S.notInverted (C.sliceV w0' vA)
+
+        aCount <- ensureLowerDensityInWindow w0' g mv hCount
+
+        go w0' (C.toVar hCount) aCount
+      where
+        -- Ensure low densities in all windows containing @w@ until the first
+        -- stability window after the intersection.
+        --
+        -- @hc@ is the number of active slots in the honest schema in @w@
+        -- @ac@ is the number of active slots in the adversarial schema in @w@
+        go w hc ac =
```

Comment:
I think the `go` loop is fine, but it might be nicer to further separate the generation of the prefixes into something like `prefixesFromTo`, which could be used like this:

```
let
    prefix = the race window
    stab   = the stability window (or the end of the mv)
(\snoc -> foldM snoc (hCount, aCount) (prefixesFromTo prefix stab)) $ \(hc, ac) -> do
    ...
    pure (hc', ac')
```

I'm guessing `prefixesFromTo` could ultimately be implemented a list comprehension.

Reply:
Adopted in 41db9fe.